### PR TITLE
Add a test runner

### PR DESCRIPTION
**Pull request by paulirish**
Added assertions for most of the functionality I saw from PRs.

### CODE REVIEW

**Comment by paulirish** [comment]
cc @ffes @davidosomething 

I've used bats before with [diff-so-fancy](https://github.com/so-fancy/diff-so-fancy) and its worked well. 

Take a look!

File: README.md [hunk 1]
```diff
@@ -138,10 +138,25 @@ from which this plugin was forked.
...
+
+```sh
+git submodule update --init # pull in the assertion libraries
+
+# Run the test suite once:
+bats test
+
+# Run it on every change with `entr`
+brew install entr
+ls --color=never * test/*.bats | entr bats test
```
Comment (davidosomething) [inline README.md]: can't recommend recommending this -- http://mywiki.wooledge.org/ParsingLs

File: test/git-open.bats [hunk 2]
```diff
@@ -0,0 +1,189 @@
...
+#   * gitopen.gitlab.protocol
+#   * Atlassian Bitbucket Server (https://github.com/paulirish/git-open/pull/15)
+
+
+teardown() {
+  cd ..
+  rm -rf "$foldername"
+}
+
+# helper to create a test git sandbox that won"t dirty the real repo
```
Comment (davidosomething) [inline test/git-open.bats]: nitpicking won"t -> won't

**Comment by davidosomething** [comment]
looks good, haven't run the tests yet for lack of time
the changes to `$open` in the actual script are fine 👍  👍

**Comment by nwinkler** [comment]
This is great! Let me know once this is merged and I'll look into adding some additional tests, e.g. for Bitbucket Server (since I'm mostly using `git open` for that).

**Comment by ffes** [comment]
Having automated tests is **always** a good thing!

Since I never had heard of bats before. I have to take a closer look to see if it works on Cygwin (my primary git client) although it seems "just a bash script" so why not.

But it is going to be a busy period so I hope I find the time.

**Comment by ffes** [comment]
Love the fact that I can test/debug with this command line:

```sh
BATS_CWD="." ./git-open
```

This should be mentioned in `README.md`.

Also tried to run `bats test` on Cygwin. It is quite slow (mainly anti virus to blame for that) but it works although 2 tests fail.

```
✗ gh: git open issue
   (from function `assert_output' in file test/test_helper/bats-assert/src/assert.bash, line 239,
    in test file test/git-open.bats, line 65)
     `assert_output "https://github.com/paulirish/git-open/issues/12"' failed
   Switched to a new branch 'master'
   Switched to a new branch 'issues/#12'

   -- output differs --
   expected : https://github.com/paulirish/git-open/issues/12
   actual   : Git remote is not set for issue
```

```
✗ gitlab seperate domains
   (from function `assert_output' in file test/test_helper/bats-assert/src/assert.bash, line 239,
    in test file test/git-open.bats, line 136)
     `assert_output "https://gitlab.example.com/namespace/project"' failed
   Switched to a new branch 'master'

   -- output differs --
   expected : https://gitlab.example.com/namespace/project
   actual   : https://git.example.com/namespace/project
   --
```

**Comment by ffes** [comment]
You can add this test as well:

```
@test "gh git protocol" {
  git remote set-url origin "git://github.com/user/repo.git"
  git checkout -B "master"
  run ../git-open
  assert_output "https://github.com/user/repo"
}
```

The current master doesn't support it, but the @derimagia rewrite handles the git protocol as well.

File: test/git-open.bats [hunk 3]
```diff
@@ -0,0 +1,189 @@
...
+  run ../git-open
+  refute_output --partial "@"
+}
+
+
+##
+## GitLab
+##
+
+@test "gitlab seperate domains" {
```
Comment (ffes) [inline test/git-open.bats]: typo: separate

File: test/git-open.bats [hunk 4]
```diff
@@ -0,0 +1,189 @@
...
+  run ../git-open
+  assert_output "https://github.com/user/repo/"
+
+  # https://github.com/paulirish/git-open/pull/30
+  git remote set-url origin "ssh://git@github.com/user/repo.git"
+  run ../git-open
+  assert_output "https://github.com/user/repo/"
+}
+
+@test "gh: git open issue" {
```
Comment (ffes) [inline test/git-open.bats]: The only name that has a ":" in the name

File: README.md (hunk 1 above)
Comment (paulirish) [inline README.md]: ah interesting. switched to `find` instead and threw into an npm script.

File: test/git-open.bats (hunk 2 above)
Comment (paulirish) [inline test/git-open.bats]: done

File: test/git-open.bats (hunk 3 above)
Comment (paulirish) [inline test/git-open.bats]: done

File: test/git-open.bats (hunk 4 above)
Comment (paulirish) [inline test/git-open.bats]: thx. i added colons to the rest of them:

image

**Comment by paulirish** [comment]
> You can add this test as well. (gh git protocol)

added.


-------

@nwinkler i'm merging now so you can followup with a bitbucket server test. :)

**Pull request merged**